### PR TITLE
feat(lua): add `vim.loader`

### PR DESCRIPTION
**Pull request by folke**
This is the [lazy.nvim](https://github.com/folke/lazy.nvim) implementation of a new lua module loader.

## Features

- **find**: fast modpath find implementation that works by caching top-level modules under `/lua` to prevent too many stat calls. The find method ensures correct rtp order.
- when `enabled`:
  - `loadfile`
  - adds the cache lua loader
  - adds the cache libs loader
  - removes the Neovim loader

## Enabling the new loader

To us the new loader, add the code below to the top of your `init.lua`:

```lua
vim.loader.enable()
```

## Interoperability with lazy.nvim

In order to increase adoption of the Neovim cache, it's important to keep
interoperability with lazy.nvim (and possibly others).

**lazy.nvim** is being used by thousands of users, so it would be a shame
if the new Neovim cache would have to be disabled when using lazy to make it work.

For this, there are a couple of low-level cache methods exposed so that lazy
can use them.

I'm planning to keep the cache implementation in sync for both lazy and Neovim.

# Exposed methods and how they are used by lazy

* `reset(path)`: needed to be called after a plugin updates or has just been installed. This will clear the top-level modules from the path's `/lua` directory.
* `find`: 
  - lazy adds a custom package loader at the very end so it can automatically load a plugin when a lua module is required from a plugin that's not been loaded yet. Here I use the `find` method with a custom set of paths (of the unloaded plugins) and then use `loadfile` which will load the module using the cache
  - lazy automatically can import plugin specs from a certain lua module. `find` is used here to find a directory for a lua module so I can list the lua files in there to load their specs.

### CODE REVIEW

**Comment by folke** [comment]
That's the idea, but I'd rather have it initially as an experimental feature where users need to opt-in to use it.

Afterwards the plan is indeed to always enable the cache and have a way to opt-out

**Comment by bfredl** [comment]
Yes, eventually this should integrate with the default loader (i.e. just be the default loader) rather than monkeypatch disable each other. Worth discussing for the timeframe of 0.10.

**Comment by lewis6991** [comment]
> What are the downsides of caching? In my (limited) experience with plugins like impatient.nvim, I occasionally had issues with changes to Lua files not being picked up when working on Neovim core files, but a simple "flush the cache" command was all that was necessary.

I think these were bugs with impatient that have since long been resolved. With impatient you only need to flush the cache when it grows too big, however as folke points out, this isn't an issue with this implementation.

File: runtime/lua/vim/cache.lua [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+  ---@type table<string, table<string,true>>
+  _indexed = {},
+  ---@type table<string, string[]>
+  _topmods = {},
+  _loadfile = loadfile,
+}
+
+--- Tracks the time spent in a function
+---@private
+function M.track(stat, start)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:41]: In addition to `@private`, can you also prefix these with `_` (underscore)?

File: runtime/lua/vim/cache.lua [hunk 2]
```diff
@@ -0,0 +1,481 @@
...
+end
+
+--- Finds the module path for the given module name
+---@param modname string
+---@param opts? CacheFindOpts (table|nil) Options for finding a module:
+---    - rtp: (boolean) Search for modname in the runtime path (defaults to `true`)
+---    - patterns: (string[]) Paterns to use (defaults to `{"/init.lua", ".lua"}`)
+---    - paths: (string[]) Extra paths to search for modname (defaults to `{}`)
+---@return string? modpath, CacheHash? hash (string|nil) modpath for the module
+function M.find(modname, opts)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:270]: Does this also check `rtp` is there is a cache miss?

File: runtime/lua/vim/cache.lua [hunk 3]
```diff
@@ -0,0 +1,481 @@
...
+--- * adds the lua loader
+--- * adds the libs loader
+--- * remove the Neovim loader
+function M.enable()
+  if M.enabled then
+    return
+  end
+  M.enabled = true
+  vim.fn.mkdir(vim.fn.fnamemodify(M.path, ':p'), 'p')
+  -- selene: allow(global_usage)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:360]: Remove.

File: runtime/lua/vim/cache.lua [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+
+--- Disables the cache:
+--- * removes the cache loaders
+--- * adds the Neovim loader
+function M.disable()
+  if not M.enabled then
+    return
+  end
+  M.enabled = false
+  -- selene: allow(global_usage)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:386]: remove

File: runtime/lua/vim/cache.lua (hunk 2 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:270]: It checks the rtp by default, unless `rtp=false`. Is that what you mean?

I use it in lazy to find modules from plugins that are not yet loaded. For that I pass `rtp=false` and `paths=paths_of_unloaded_plugins`

File: runtime/lua/vim/cache.lua [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+      table.remove(package.loaders, l)
+    end
+  end
+  table.insert(package.loaders, 2, vim._load_package)
+end
+
+--- Return the top-level `/lua/*` modules for this path
+---@param path string path to check for top-level lua modules
+---@return string[]
+function M.lsmod(path)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:400]: Can this be made private?

File: runtime/lua/vim/cache.lua [hunk 6]
```diff
@@ -0,0 +1,481 @@
...
+        end
+      end
+    end
+    M.track('lsmod', start)
+  end
+  return Cache._indexed[path]
+end
+
+--- Debug function that wrapps all loaders and tracks stats
+function M.profile_loaders()
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:436]: Maybe make this private too?

File: runtime/lua/vim/cache.lua [hunk 7]
```diff
@@ -0,0 +1,481 @@
...
+      local ret = loader(modname)
+      M.track('loader ' .. l .. ': ' .. loc, start)
+      M.track('loader_all', start)
+      return ret
+    end
+  end
+end
+
+--- Prints all cache stats
+function M.inspect()
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:450]: And this. This should really be just for developer use.

File: runtime/lua/vim/cache.lua [hunk 8]
```diff
@@ -0,0 +1,481 @@
...
+    -- trigger correct error
+    chunk, err = Cache._loadfile(modpath, opts.mode, opts.env)
+    M.track('load', start)
+    return chunk, err
+  end
+
+  local entry = Cache.read(modpath)
+  if entry and Cache.eq(entry.hash, hash) then
+    -- found in cache and up to date
+    -- selene: allow(incorrect_standard_library_use)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:245]: remove

File: runtime/lua/vim/cache.lua [hunk 9]
```diff
@@ -0,0 +1,481 @@
...
+
+--- Loads the given module path using the cache
+---@param modpath string
+---@param opts? {hash?: CacheHash, mode?: "b"|"t"|"bt", env?:table} (table|nil) Options for loading the module:
+---    - hash: (table) the hash of the file to load if it is already known. (defaults to `vim.loop.fs_stat({modpath})`)
+---    - mode: (string) the mode to load the module with. "b"|"t"|"bt" (defaults to `nil`)
+---    - env: (table) the environment to load the module in. (defaults to `nil`)
+---@see |luaL_loadfile()|
+---@return function?, string? error_message
+function M.load(modpath, opts)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:227]: Private?

**Review by lewis6991 — COMMENTED** [review]
I'm happy if you want to expose a bunch of functions in an unstable way, but really do not like providing (and documenting) all the low level functions.

**Comment by justinmk** [comment]
> In order to increase adoption ... keep interoperability with lazy.nvim
> ... For this, there are a couple of low-level cache methods exposed so that lazy can use them.

Are these low-level things permanently needed by lazy.nvim or is this a temporary "bridging" step?

If plugin (managers) need these low-level functions to effectively use `vim.cache`, then they're a permanent API, right?

**Comment by clason** [comment]
> That's the idea, but I'd rather have it initially as an experimental feature where users need to opt-in to use it.

It makes sense to put this as opt-in "early access" for 0.9 so devs can dog-food over the 0.10 cycle (like we did with Lua treesitter highlighting), but I agree the goal should be to make it work seamless enough so that it can be opt-out (e.g.., via `--luamod-dev` or similar, as I don't see why you'd _never_ want to have caching).

File: runtime/lua/vim/cache.lua [hunk 10]
```diff
@@ -0,0 +1,481 @@
+-- interop with the native Neovim cache
+if type(package.loaded['vim.cache']) == 'table' then
+  return package.loaded['vim.cache']
+end
+
+local ffi = require('ffi')
```
Comment (mfussenegger) [inline runtime/lua/vim/cache.lua:6]: This means this only works with luajit, right?

File: runtime/lua/vim/cache.lua (hunk 7 above)
Comment (justinmk) [reply runtime/lua/vim/cache.lua:450]: we will want to have visibility  `:checkhealth` , somehow

File: runtime/lua/vim/cache.lua (hunk 10 above)
Comment (lewis6991) [reply runtime/lua/vim/cache.lua:6]: Yes, we need to remove this.

Comment (clason) [reply runtime/lua/vim/cache.lua:6]: Or disable the cache for PUC Lua builds? If you build against that, you obviously do not care about speed at all...

Comment (lewis6991) [reply runtime/lua/vim/cache.lua:6]: It's not needed though.

Comment (lewis6991) [reply runtime/lua/vim/cache.lua:6]: Since Lua 5.1 is a strictly defined language and LuaJIT is just an implementation, I'm quite keen on the idea of us being able to move across to any 5.1 implementation that comes about.

File: runtime/lua/vim/cache.lua (hunk 7 above)
Comment (clason) [reply runtime/lua/vim/cache.lua:450]: Should probably be something separate, though; this is more like low-level debug information reporting the timings of various cache loader steps. `:checkhealth` should be more high-level ("cache is enabled and has size x Kb").

File: runtime/lua/vim/cache.lua (hunk 10 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:6]: @lewis6991 i wouldn't use mpack for the file-based cache.

I think it's better to just not enable the cache when `ffi` is not available. Will change it.

Comment (lewis6991) [reply runtime/lua/vim/cache.lua:6]: > i wouldn't use mpack for the file-based cache.

Is there a technical reason why?

Comment (folke) [reply runtime/lua/vim/cache.lua:6]: Alright, nevermind. Will change the implementation so it no longer needs ffi and also won't need mpack

File: runtime/lua/vim/cache.lua
[the remark this answers is not in the document]
Comment (folke) [reply runtime/lua/vim/cache.lua:1]: nice! Forgot that was added. Will add a short decsription

**Comment by folke** [comment]
> Are these low-level things permanently needed by lazy.nvim or is this a temporary "bridging" step?
> 
> If plugin (managers) need these low-level functions to effectively use `vim.cache`, then they're a permanent API, right? Is `vim.cache` useful without those functions?

Those methods will always be needed by lazy, so I'd really would like those to be kept permanent indeed.

In the code, all the methods returned on the module are used. The ones on the `Cache` object are only internally used by the cache, so dont need to be exposed.

I've updated the PR description with where lazy uses those.

**Comment by lewis6991** [comment]
> reset(path): needed to be called after a plugin updates or has just been installed. This will clear the top-level modules from the path's /lua directory.

Shouldn't the cache to invalidating this automatically?

File: runtime/lua/vim/cache.lua [hunk 11]
```diff
@@ -0,0 +1,481 @@
...
+---@class ModuleCache
+---@field _rtp string[]
+---@field _rtp_pure string[]
+---@field _rtp_key string
+local Cache = {
+  ---@type table<string, table<string,true>>
+  _indexed = {},
+  ---@type table<string, string[]>
+  _topmods = {},
+  _loadfile = loadfile,
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:36]: Why is this here? Can't `Cache.loadfile` be use throughout?

File: runtime/lua/vim/cache.lua [hunk 12]
```diff
@@ -0,0 +1,481 @@
...
+  local start = uv.hrtime()
+
+  opts = opts or {}
+  local hash = opts.hash or uv.fs_stat(modpath)
+  ---@type function?, string?
+  local chunk, err
+
+  if not hash then
+    -- trigger correct error
+    chunk, err = Cache._loadfile(modpath, opts.mode, opts.env)
```
Comment (lewis6991) [inline runtime/lua/vim/cache.lua:237]: `Cache._loadfile` also calls `M.load`? How does that get resolved?

File: runtime/lua/vim/cache.lua (hunk 11 above)
Comment (lewis6991) [reply runtime/lua/vim/cache.lua:36]: Sorry, I see `Cache._loadfile` is just `_G.loadfile`. Can this be pulled out?

File: runtime/lua/vim/cache.lua (hunk 12 above)
Comment (lewis6991) [reply runtime/lua/vim/cache.lua:237]: I get it now.

File: runtime/lua/vim/cache.lua (hunk 11 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:36]: why? I override _G.loadfile with the Cache.loadfile implementation, but I of course still need the original function to load a file when needed

File: runtime/lua/vim/cache.lua (hunk 7 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:450]: I'd rather keep this, since it's similar to other inspect methods we have?

File: runtime/lua/vim/cache.lua (hunk 10 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:6]: @lewis6991 performance would be greatly impacted. But I can prevent using either. Working on a better solution.

File: runtime/lua/vim/cache.lua (hunk 7 above)
Comment (clason) [reply runtime/lua/vim/cache.lua:450]: I was referring to `:checkhealth` specifically; it should not use this `inspect` but something else better suited to that context.

No opinion on `inspect()` itself.

**Comment by folke** [comment]
> Shouldn't the cache to invalidating this automatically?

I have no control when lua files and directories are being added while Neovim is open to paths on the rtp.

Invalidation of loading the same file twice is of course automatic.

The reset is used in the cases where a new top-level folder was added to a path on the rtp's `/lua` directory.

Not likely in normal usage, but of course very likely for a plugin manager like lazy while installing and updating plugins.

File: runtime/lua/vim/cache.lua (hunk 11 above)
Comment (lewis6991) [reply runtime/lua/vim/cache.lua:36]: I mean just do something like: `local orig_loadfile = _G.loadfile`. There's no need to place this in a table.

Comment (folke) [reply runtime/lua/vim/cache.lua:36]: there's also no harm in it? The whole Cache module is private anyway

File: runtime/lua/vim/cache.lua (hunk 10 above)
Comment (lewis6991) [reply runtime/lua/vim/cache.lua:6]: How? It's simply used to read/write a single header.

How greatly are we talking?

**Comment by lewis6991** [comment]
> I have no control when lua files and directories are being added while Neovim is open to paths on the rtp.

No, but you need to check the file exists and matches the hash when loading a file. Therefore, you should never need to reset. This is the premise on which I suggested adding the cache to core to begin with.

**Comment by folke** [comment]
> No, but you need to check the file exists and matches the hash when loading a file. Therefore, you should never need to reset. This is the premise on which I suggested adding the cache to core to begin with.

As I said, that is of course automatic.

The reset is only needed when a folder on the rtp has new top-level modules all of the sudden, within the same Neovim session.

Suppose you have a folder X:
`lua/x/a.lua`
`lua/y/a.lua`

When you do `require("x.a")`, the top-level mods will be loaded since we haven't loaded them yet, so the top-level mods for folder `X` will be `{"x", "y"}`

If you then do `require("y.a")`, top-level mods for `X` are already known, so we will know that rtp folder `X` is a possible candidate to check for the `/lua/y/a.lua` file.

Suppose that something then adds `X/lua/z/a.lua`, and does `require("z.a")`, then the top-level mods for `X` will still be `x,y`.

This is something that doesn't happen in normal usage.

But again, for a plugin manager like lazy, this can happen during updates/installs, so I clear the top-level mods for `X` so that when a new `require` happens, the top-level mods will be re-loaded.

**Comment by lewis6991** [comment]
> Suppose that something then adds X/lua/z/a.lua, and does require("z.a"), then the top-level mods for X will still be x,y.

Then the cache lookup fails, and loads the module as normal and updates the cache.

File: runtime/lua/vim/cache.lua (hunk 10 above)
Comment (jamessan) [reply runtime/lua/vim/cache.lua:6]: > except to show PUC Lua builders the error of their ways :]

Yes, how dare people run nvim on platforms that LuaJIT doesn't support!

Comment (clason) [reply runtime/lua/vim/cache.lua:6]: Exactly!

File: runtime/lua/vim/cache.lua (hunk 11 above)
Comment (lewis6991) [reply runtime/lua/vim/cache.lua:36]: Ok, leave it for now then. I can go over it all at a later point.

**Comment by folke** [comment]
> Then the cache lookup fails, and loads the module as normal and updates the cache.

and that would make all the performance gains moot, since unfortunately a lot of plugins actually try requiring modules that don't exist.

Every time that happens, we'd then have to clear all the topmods for every path on the rtp and re-calculate them.

btw, I'm pretty sure that getruntimefile with `is_lua` has the same problem. (adding a path on the rtp with `/lua`, then add `/lua/something.lua` and the new mod won't be found until  you restart Neovim)

File: runtime/lua/vim/cache.lua (hunk 7 above)
Comment (justinmk) [reply runtime/lua/vim/cache.lua:450]: > I'd rather keep this, since it's similar to other inspect methods we have?

+1 but mark it as private and/or "experimental" so that we don't have to think much about future-proofing it yet? I assume that was @lewis6991 's main concern (which I agree with).

**Comment by lewis6991** [comment]
> and that would make all the performance gains moot, since unfortunately a lot of plugins actually try requiring modules that don't exist.

I just don't think this is true. I'm aware of the `pcall(require, 'noexist')` issue, and typically this does not happen often enough to degrade performance.

**Comment by folke** [comment]
@lewis6991 
image

And with your suggested changes:

image

**Comment by lewis6991** [comment]
Is this because lazy or your config does lots of `require('noexist')`?

**Comment by justinmk** [comment]
> How can you know where the miss was? A non-existent module could have been added to any path on the rtp

We could track failed require() with `verbose=1` (`nvim -V1`, which we have established as the "instrumentation mode". Then `:checkhealth` could report the call paths.

That adds a bit of sleuthing for users trying to max out their performance, but that will already be the case anyway, `reset()` doesn't really avoid that.

**Comment by folke** [comment]
It's because of some plugins, like mason for example.

File: runtime/lua/vim/cache.lua (hunk 10 above)
Comment (folke) [reply runtime/lua/vim/cache.lua]: Just changed the implementation so it no longer needs ffi (or mpack). Performance is the same.

**Comment by folke** [comment]
> A fair good point. I'm sure there's still a way to work around this.

How? How could there be a way to work around this?

Again, this is only an issue when a new top-level mod was added to any of the paths on the rtp. You can never know where it was added. Using luv watch for this purpose could detect that, but I don't think that's a great idea since you'd have to monitor all the rtp paths.

Always clearing the top-mods when a module wasn't found, completely trashes any performance gains of using the cache.

And as I mentioned, the existing Neovim loader also has a similar issue. When you create a folder without `/lua` and add that to the rtp. try doing `require("foobar")`. it will fail. Now create `/lua/foobar.lua` in that new folder. and do the require again. It will still fail. Exact same issue.

File: runtime/lua/vim/cache.lua (hunk 5 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:397]: used by lazy, so I don't think it's a good idea to make this private

File: runtime/lua/vim/cache.lua (hunk 6 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:433]: used by lazy, so I don't think it's a good idea to make this private

File: runtime/lua/vim/cache.lua (hunk 9 above)
Comment (folke) [reply runtime/lua/vim/cache.lua:226]: used by lazy, so I don't think it's a good idea to make this private

**Comment by folke** [comment]
@lewis6991 I just added the autocmd below to the cache, so when a user creates a new top-level module during a NEovim session for a path that is on the rtp, then the top-mods will be reset automatically and requiring those new modules would work without restart.

The only way this could still fail is when during a Neovim session new files/directories have been added to directories on the rtp. Like by lazy.nvim for example, so the low-level `reset()` function is still needed for those cases.

```lua
  -- this will reset the top-mods in case someone adds a new
  -- top-level lua module to a path already on the rtp
  vim.api.nvim_create_autocmd("BufWritePost", {
    group = vim.api.nvim_create_augroup("cache_topmods_reset", { clear = true }),
    callback = function(event)
      local bufname = event.match ---@type string
      local idx = bufname:find("/lua/", 1, true)
      if idx then
        M.reset(bufname:sub(1, idx - 1))
      end
    end,
  })
```

**Comment by folke** [comment]
@justinmk I'm fine in making `inspect` and `profile_loaders` `@private`, since those methods aren't used by lazy.

But what should I do with the other low-level methods that I use (the ones explained in the PR description)?

I can make those private too, but not sure this is a good idea, since lazy (and as such thousands of users) depends on them.

Wdyt?

File: runtime/lua/vim/cache.lua [hunk 13]
```diff
@@ -0,0 +1,489 @@
...
+  return '\ncache_loader: module ' .. modname .. ' not found'
+end
+
+--- The `package.loaders` loader for libs
+---@param modname string module name
+---@return string|function
+---@private
+function Cache.loader_lib(modname)
+  local start = uv.hrtime()
+  local modpath = M.find(modname, { patterns = jit.os:find('Windows') and { '.dll' } or { '.so' } })
```
Comment (mfussenegger) [inline runtime/lua/vim/cache.lua:169]: `jit.os` is luajit only as well?

**Review by mfussenegger — COMMENTED** [review]
Other random thought: is `vim.cache` the right namespace for this? When I saw the PR title I was thinking this would add a general purpose cache structure (table with eviction policy,  max size,  etc.)

**Comment by clason** [comment]
# Current plan

This is a very welcome feature we definitely would like to have in core; as the point is improving performance, optimizations like the ones (passionately) discussed above should be fair game.

Nevertheless, we need to spend some time thinking how we can shape the API surface so that it fits within the general design of the Neovim API while still allowing plugins like Lazy.nvim (and hopefully many others!) to make use of it without sacrificing performance.

We'll therefore revisit this after the 0.9 release with the goal of having a stable API for Lazy for 0.10.

File: runtime/lua/vim/loader.lua [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+
+local M = {}
+
+---@alias CacheHash {mtime: {sec:number, nsec:number}, size:number}
+---@alias CacheEntry {hash:CacheHash, chunk:string}
+
+---@class ModuleFindOpts
+---@field all? boolean Search for all matches (defaults to `false`)
+---@field rtp? boolean Search for modname in the runtime path (defaults to `true`)
+---@field patterns? string[] Paterns to use (defaults to `{"/init.lua", ".lua"}`)
```
Comment (gpanders) [inline runtime/lua/vim/loader.lua:11]: Typo here. Also does "patterns" here mean "Lua patterns"? Given that Vim has its own "pattern" concept, it might be good to be explicit.

```suggestion
---@field patterns? string[] Lua patterns to use (defaults to `{"/init.lua", ".lua"}`)
```

Comment (folke) [reply runtime/lua/vim/loader.lua:11]: Not lua patterns and also not vim patterns.

Comment (gpanders) [reply runtime/lua/vim/loader.lua:11]: So glob patterns then? Or just plain strings? The term "patterns" is ambiguous in Vim precisely because of how many different pattern "schemes" are in use.

Comment (folke) [reply runtime/lua/vim/loader.lua:11]: no, just plain strings.

Comment (clason) [reply runtime/lua/vim/loader.lua:11]: In the Lua context, though, it's usually (and should always be) used for "Lua patterns" (aka "we already have regex at home").

Comment (folke) [reply runtime/lua/vim/loader.lua:11]: Those aren't lua patterns. They're just strings that are added to the basename of the lua module.

Comment (folke) [reply runtime/lua/vim/loader.lua:11]: What should I rename it to? Any suggestions?

Comment (gpanders) [reply runtime/lua/vim/loader.lua:11]: The name itself is fine to me, as long as the docstring is explicit about the semantics.

>They're just strings that are added to the basename of the lua module.

This is useful information to include in the docstring. As a user of this API this is exactly what I would want to know.

File: runtime/lua/vim/loader.lua
Comment (lewis6991) [inline runtime/lua/vim/loader.lua:1]: Would it be ok to localize all non-documented functions? E.g. `_track`, `normalize`, `Loader.get_rtp`, etc.

File: runtime/lua/vim/loader.lua [hunk 15]
```diff
@@ -0,0 +1,528 @@
...
+  _indexed = {},
+  ---@type table<string, string[]>
+  _topmods = {},
+  _loadfile = loadfile,
+}
+
+--- Tracks the time spent in a function
+---@private
+function M._track(stat, start)
+  M.stats[stat] = M.stats[stat] or { total = 0, time = 0 }
```
Comment (justinmk) [inline runtime/lua/vim/loader.lua:42]: should `M.stats`  be local to the module rather than on `M` itself?

**Review by justinmk — APPROVED** [review]
I think we're aligned on having this for 0.9 so that plugins can start using it. So, if we want to merge this for that reason, let's do so. But while it's top-of-mind, we will soon need:

1. troubleshooting steps written down somewhere
2. test coverage

File: runtime/lua/vim/loader.lua (hunk 15 above)
Comment (folke) [reply runtime/lua/vim/loader.lua:42]: I currently show those stats in the Lazy debug ui, but that's not that important.

I'll localize `stats` and will for now add it to `M._inspect()`, which is already private. If that method would be removed at some point, then no big deal. I'll then just remove the debug output from the lazy ui.

File: runtime/lua/vim/loader.lua
Comment (folke) [reply runtime/lua/vim/loader.lua:1]: Actually, it's probably a good idea to keep `M._Loader = Loader`? Useful for writing tests (that I still need to port) and later for debugging.

Comment (lewis6991) [reply runtime/lua/vim/loader.lua:1]: Sure.

Comment (folke) [reply runtime/lua/vim/loader.lua:1]: done

File: runtime/lua/vim/loader.lua (hunk 15 above)
Comment (folke) [reply runtime/lua/vim/loader.lua:42]: Made stats local and changed the private `M._inspect` method to return the raw stats and optionally print them

**Comment by folke** [comment]
I have a bunch of tests for the loader in lazy that I can port. Will see how can integrate them tomorrow. There's some tricky parts since I need to create a dummy filesystem, but the Neovim testing framework, probably already has helpers for that. Will look into it.

**Review by lewis6991 — APPROVED** [review]

**Comment by lewis6991** [comment]
Since this is opt-in we can defer adding tests until later.
